Generate custom-icon-webfont on each provision or update-prod-static.

rht opened the pull request:
RFCR @pweaver. I have made sure that references in `zproject/settings.py` and elsewhere, checked via grep.

rht commented:
I have tested with doing a `./tools/build-release-tarball` and gut out its innard:
```
(zulip-py3-venv) ~/rht_zulip/foo/zulip-server-webfont$ ls prod-static/serve/generated/icons/fonts
zulip-icons.01941e73afa6.eot  zulip-icons.b61502b7f02b.svg    zulip-icons.de770e4232f0.woff  zulip-icons.svg  zulip-icons.woff
zulip-icons.03d466915d50.ttf  zulip-icons.d36be9e1624b.woff2  zulip-icons.eot                zulip-icons.ttf  zulip-icons.woff2
```

rht commented:
Got this error when trying to `upgrade-zulip-from-git` to this PR:
```
Traceback (most recent call last):
  File "./tools/update-prod-static", line 53, in <module>
    stdout=fp, stderr=fp)
  File "/usr/lib/python3.4/subprocess.py", line 561, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./tools/generate-custom-icon-webfont']' returned non-zero exit status 7
Traceback (most recent call last):
  File "/home/zulip/deployments/2017-11-16-11-31-29/scripts/lib/upgrade-zulip-stage-2", line 82, in <module>
    preexec_fn=su_to_zulip)
  File "/usr/lib/python3.4/subprocess.py", line 561, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./tools/update-prod-static', '--authors-not-required', '--prev-deploy', '/home/zulip/deployments/current']' returned non-zero exit status 1
Traceback (most recent call last):
  File "/home/zulip/deployments/current/scripts/upgrade-zulip-from-git", line 67, in <module>
    deploy_path, "--from-git"] + deploy_options)
  File "/usr/lib/python3.4/subprocess.py", line 561, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sudo', '/home/zulip/deployments/2017-11-16-11-31-29/scripts/lib/upgrade-zulip-stage-2', '/home/zulip/deployments/2017-11-16-11-31-29', '--from-git']' returned non-zero exit status 1
```

No logfile in `/var/log/zulip/upgrade.log`. I suspect this is due to permission problem.

rht commented:
One hypothesis of the error: `upgrade-zulip-from-git` uses the old version of package.json without the `webfonts-generator` in production dependencies, and so `./tools/generate-custom-icon-webfont` fails to run.

I suppose the only way to fix it is to add `webfonts-generator` first and vendor it into release, then upgrade from that release into this one?

pweaver commented:
Probably not. `upgrade-zulip-from-git` definitely uses the new `package.json`. If there are permission errors creating the log file, the might itself be enough to crash the upgrade script.

pweaver commented on tools/generate-custom-icon-webfont:
```diff
@@ -34,6 +34,6 @@ fs.readdir('./static/assets/icons', function (err, files) {
             process.exit(1);
         }
-        console.log('SUCCESS! Webfonts generated in /static/icons/fonts.');
+        console.log('SUCCESS! Webfonts generated in /static/generated/icons/fonts.');
```
Should be ./static/generated/icons/fonts

rht replied:
Addressed.

timabbott commented:
This doesn't update yarn.lock; that's probably the issue with dependencies, right?

pweaver commented:
@timabbott I don't think so? I don't think yarn.lock cares if dependencies are in `dependencies` or dev dependencies. It should generate an identical file if they move from one to another.

@rht I think you can also remove the `webfont-generator` from `devDependencies`. `devDependencies` are additional dependencies on a dev install.

pweaver commented:
Also, it looks like yarn.lock hasn't been updated in a while and an update should probably be pushed to master.

timabbott commented:
@pweaver sounds good; remind me the procedure for that?

pweaver commented:
Run `yarn install` in the top level directory and it should update the `yarn.lock` which can be committed.

rht commented:
> @rht I think you can also remove the webfont-generator from devDependencies. devDependencies are additional dependencies on a dev install.

Addressed.

> Run yarn install in the top level directory and it should update the yarn.lock which can be committed.

Addressed as well.

rht commented:
Updating the `yarn.lock` via `yarn install` fixed the `upgrade-zulip-from-git` issue for me.
```
2017-11-17 00:09:18,374 upgrade-zulip-stage-2: Upgrade complete!
Cleaning unused deployments...
Deployments cleaned successfully...
Cleaning orphaned/unused caches...
Cleaning unused venv caches...
Cleaning unused node modules caches...
Cleaning unused emoji caches...
Done!

```

timabbott commented:
This looks good, but you'll need to rebase (I just merged some other changes to `yarn.lock`).

rht commented:
Rebased.

timabbott commented:
Merged, after updating PROVISION_VERSION for the merge conflict, thanks @rht!

@synicalsyntax FYI; I'd appreciate a post-commit review from you on this :).
